File: src/test/js/formulas/math/ACOTTest.ts
import { assert } from "chai";
import { it, describe } from "../../testutils/TestUtils";
import { stub } from "sinon";
import { F7ExceptionName } from "../../../../main/js/errors/F7ExceptionName";
import { NAException } from "../../../../main/js/errors/NAException";
import { ValueException } from "../../../../main/js/errors/ValueException";
import { ACOT } from "../../../../main/js/formulas/math/ACOT";
import { Grid } from "../../../../main/js/models/common/Grid";
import { CollateralLookupFunction, LookupFunction } from "../../../../main/js/models/common/Types";
import { CommonModels } from "../../testutils/CommonModels";

describe("ACOT", function () {
  it("should work with numbers", function () {
    assert.deepEqual(ACOT.SELF.run(null, 10), 0.09966865249116204);
    assert.deepEqual(ACOT.SELF.run(null, 128731.2), 7.768124588133144e-6);
    assert.deepEqual(ACOT.SELF.run(null, 11.11), 0.08976710276137885);
    assert.deepEqual(ACOT.SELF.run(null, 0), 1.570796327);
    assert.deepEqual(ACOT.SELF.run(null, 88281), 1.1327465705613094e-5);
    assert.deepEqual(ACOT.SELF.run(null, 2), 0.4636476090008061);
    assert.deepEqual(ACOT.SELF.run(null, 4), 0.24497866312686414);
    assert.deepEqual(ACOT.SELF.run(null, -4), -0.24497866312686414);
    assert.deepEqual(ACOT.SELF.run(null, -10124), -9.877518735162196e-5);
  });

  it("should work with strings", function () {
    assert.equal(ACOT.SELF.run(null, "10"), 0.09966865249116204);
  });

  it("should do pass-through errors", function () {
    assert.deepEqual(ACOT.SELF.run(null, new ValueException()), new ValueException());
  });

  it("should use lookup", function () {
    const lookup = stub();
    const collateralLookup = stub();
    const F = new ACOT(lookup as LookupFunction, collateralLookup as CollateralLookupFunction);
    collateralLookup.withArgs(CommonModels.A1, CommonModels.M22_RANGE).returns(10);
    assert.deepEqual(F.run(CommonModels.A1, CommonModels.M22_RANGE), 0.09966865249116204);
    assert.isTrue(collateralLookup.calledWith(CommonModels.A1, CommonModels.M22_RANGE));
    assert.equal(collateralLookup.callCount, 1);
    assert.isTrue(lookup.notCalled);
  });

  it("should handle grids", function () {
    const one = Grid.builder().add(0, 0, 10).add(0, 1, "A").build();
    assert.deepEqual(ACOT.SELF.run(null, one), 0.09966865249116204);
  });

  it("should return error when argument lengths are wrong", function () {
    assert.deepEqual((ACOT.SELF.run(null) as NAException).name, F7ExceptionName.NA);
    assert.deepEqual((ACOT.SELF.run(null, "A", "B") as NAException).name, F7ExceptionName.NA);
  });
});

File: src/test/java/io/protobase/f7/formulas/math/ACOTTest.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.NAException;
import io.protobase.f7.errors.ValueException;
import io.protobase.f7.models.Grid;
import io.protobase.f7.testutils.TestFormula;
import org.junit.Test;

import static com.google.common.truth.Truth.assertThat;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoMoreInteractions;
import static org.mockito.Mockito.when;

public class ACOTTest extends TestFormula {
  @Test
  public void testApply() {
    assertThat(ACOT.SELF.apply(null, 10.0)).isEqualTo(0.09966865249116204);
    assertThat(ACOT.SELF.apply(null, 128731.2)).isEqualTo(7.768124588133144E-6);
    assertThat(ACOT.SELF.apply(null, 11.11)).isEqualTo(0.08976710276137885);
    assertThat(ACOT.SELF.apply(null, 0.0)).isEqualTo(1.570796327);
    assertThat(ACOT.SELF.apply(null, 88281.0)).isEqualTo(1.1327465705613094E-5);
    assertThat(ACOT.SELF.apply(null, 2.0)).isEqualTo(0.4636476090008061);
    assertThat(ACOT.SELF.apply(null, 4.0)).isEqualTo(0.24497866312686414);
    assertThat(ACOT.SELF.apply(null, -4.0)).isEqualTo(-0.24497866312686414);
    assertThat(ACOT.SELF.apply(null, -10124.0)).isEqualTo(-9.877518735162196E-5);
  }

  @Test
  public void testApply_stringConversion() {
    assertThat(ACOT.SELF.apply(null, "88281")).isEqualTo(1.1327465705613094E-5);
  }

  @Test
  public void testApply_grid() {
    assertThat(ACOT.SELF.apply(null,
        Grid.builder().add(0, 0, 88281.0).add(0, 1, "Don't mind me.").build()
    )).isEqualTo(1.1327465705613094E-5);
  }

  @Test
  public void test_lookup() {
    ACOT F = new ACOT(lookup, collateralLookup);
    when(collateralLookup.apply(A1, M22_RANGE)).thenReturn(10.0);
    assertThat(F.apply(A1, M22_RANGE)).isEqualTo(0.09966865249116204);
    verify(collateralLookup).apply(A1, M22_RANGE);
    verifyNoMoreInteractions(collateralLookup);
  }

  @Test
  public void testApply_errorsPassThrough() {
    assertThat(ACOT.SELF.apply(null, new ValueException())).isEqualTo(new ValueException());
  }

  @Test
  public void testApply_argumentsMismatch() {
    assertThat(ACOT.SELF.apply(null, "A", "Too many")).isEqualTo(new NAException());
  }
}
